pkg/util/maps/maps.go

```diff
@@ -35,24 +35,48 @@ func Clone[K comparable, V any, S ~map[K]V](s S) S {
 	return ret
 }
 
-// Merge merges a and b while resolving the conflicts by calling resolveConflict
-func Merge[K comparable, V any, S ~map[K]V](a, b S, resolveConflict func(a, b V) V) S {
+func join[K comparable, V any, M ~map[K]V](a, b M, commonKeyVal func(a, b V) V, exclusive bool) M {
 	if a == nil {
+		if exclusive {
+			return nil
+		}
 		return Clone(b)
 	}
 
-	ret := Clone(a)
+	if b == nil {
+		if exclusive {
+			return nil
+		}
+		return Clone(a)
+	}
+
+	var ret M
+	if exclusive {
+		ret = make(M, len(a))
```

Comment:
Changed the intersection allocation size to min(). Created #1047 to revisit the implementation once we switch to go 1.21.

(computing the intersection len is a bit overkill in my opinion and adds with it an additional full map scan)

Reply:
In that case, could we not preallocate, just grow from zeo? The wasted memory feels wrong for a utility method. In extreme case, when the sets are disjoint, and equal size this might be a lot.

Reply:
Looking at the apimachinery implementation of set intersection they grow the intersection from 0: https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/set.go#L161-L177. Note that the set is implemented underneath as a map.

Reply:
Starting with:

1. Not knowing the content of the tow maps and allocating the smallest length will lead in allocating double the needed size, on average.
2. The map size extension, is not done in increments of 1 which is good for runtime consumption but will add some memory waste.  

Punting all of these in some benchmarks like:
```go
func BenchmarkPrealloc(b *testing.B) {
	cases := []struct {
		name string
		f    func(l int)
	}{
		{
			name: "empty",
			f: func(l int) {
				m := map[int]int{}
				for i := 0; i < l; i++ {
					m[i] = i
				}
			},
		},
		{
			name: "makeZero",
			f: func(l int) {
				m := make(map[int]int, 0)
				for i := 0; i < l; i++ {
					m[i] = i
				}
			},
		},
		{
			name: "preallocFix",
			f: func(l int) {
				m := make(map[int]int, l)
				for i := 0; i < l; i++ {
					m[i] = i
				}
			},
		},
		{
			name: "preallocDouble",
			f: func(l int) {
				m := make(map[int]int, l*2)
				for i := 0; i < l; i++ {
					m[i] = i
				}
			},
		},
	}

	for _, tc := range cases {
		b.Run(tc.name, func(b *testing.B) {
			tc.f(b.N)
		})
	}
}
```

and running it: 

```sh
go test -bench=. -benchmem -benchtime 2s
goos: linux
goarch: amd64
pkg: loc.al/mapbm
cpu: Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz
BenchmarkPrealloc/empty-12         	19361432	       146.9 ns/op	      73 B/op	       0 allocs/op
BenchmarkPrealloc/makeZero-12      	20865111	       141.6 ns/op	      68 B/op	       0 allocs/op
BenchmarkPrealloc/preallocFix-12   	40684652	        71.81 ns/op	      32 B/op	       0 allocs/op
BenchmarkPrealloc/preallocDouble-12         	39483420	        57.30 ns/op	      65 B/op	       0 allocs/op
```
Shows that the overall memory consumption  is worst in double pre-allocation that in a fix pre-allocation but very similar to no pre-allocation, however the runtime performance is 2-3x better for double pre-allocation that no pre-allocation cases.

Reply:
Yeah, this is sort of expected. As you point out, there are tradeoffs. In that case I was suggesting to follow how other libraries do it. I'm ok keeping as is, but I would like to keep the discussion open in case @alculquicondor has some view.

Reply:
Please also note that the benchmarks omit the point from https://github.com/kubernetes-sigs/kueue/pull/1039#discussion_r1288051356, that currently the overallocation might not be 2x, but arbitrary large, when the two maps have small intersection, but are large. 

I think this is the main reason why the referenced impl grows from zero. 

And, this is probably not a problem for the current use in Kueue, but this is a utility function, so it makes sense to investigate the risks, as the function might be used elsewhere in the future.

Reply:
> Please also note that the benchmarks omit the point from [#1039 (comment)](https://github.com/kubernetes-sigs/kueue/pull/1039#discussion_r1288051356), that currently the overallocation might not be 2x, but arbitrary large, when the two maps have small intersection, but are large.

It is probable to have two large sets with a narrow intersection, the same way it is probable to have small or large sets that fully intersect. 

Regarding https://github.com/kubernetes-sigs/kueue/pull/1039#discussion_r1288051356 , it should be noted that the storage extension of the map is not done in increments of one and the "waste" is still there (maybe smaller in some cases but achieved with multiple reallocations).

Anyway with the purpose of moving this forward I'll drop the preallocation.